### ISSUE TITLE
uninstall should remove activated packages

### ISSUE THREAD

**Issue by svenevs** [LLNL/spack#3965]
**Edit**.  Thank the gods for `grep -r`.  In my rampant testing of `flann`, I uninstalled versions that had already been `activate`d.  They remained in the respective `extensions.yaml` file of my python installations.  This means that any `activate` call will break for all.

The solution would be to `deactivate` by force when `uninstall` is performed, or maybe give a warning?  Feel free to close, I don't know if this is already known but probably is.  I thought the error was something else.

```console
sven:/opt/spack> mv var/spack/repos/builtin/packages/flann/ ~/Desktop/
sven:/opt/spack> spack activate /4ufq46n
==> Error: Spec b5rjo5qif5gs44ymukihkgmy2innvwcf not found in /opt/spack/opt/spack/linux-fedora23-x86_64/gcc-5.3.1/flann-1.9.1-b5rjo5qif5gs44ymukihkgmy2innvwcf
sven:/opt/spack> spack reindex
sven:/opt/spack> spack find -lv py-numpy
==> 2 installed packages.
-- linux-fedora23-x86_64 / gcc@5.3.1 ----------------------------
4ufq46n py-numpy@1.12.1+blas+lapack  fgzexrq py-numpy@1.12.1+blas+lapack
sven:/opt/spack> spack info flann
==> Error: Package flann not found.
sven:/opt/spack> spack activate /4ufq46n
==> Error: Spec b5rjo5qif5gs44ymukihkgmy2innvwcf not found in /opt/spack/opt/spack/linux-fedora23-x86_64/gcc-5.3.1/flann-1.9.1-b5rjo5qif5gs44ymukihkgmy2innvwcf
sven:/opt/spack> grep 'flann' opt/spack/.spack-db/
index.json   lock         prefix_lock  
sven:/opt/spack> grep 'flann' opt/spack/.spack-db/index.json 
sven:/opt/spack> echo ":'("
:'(
```

**How does activate know about `flann`**?  That particular `py-numpy` came from a `flann` installation, but I moved the `flann` package out of the working tree. [i'm packaging flann right now, removed links since no longer relevant]

**Comment by tgamblin** [comment]
So, we do actually deactivate on uninstall; see `hooks/extensions.py`:

```python
def pre_uninstall(pkg):
    assert(pkg.spec.concrete)

    if pkg.is_extension:
        if pkg.activated:
            pkg.do_deactivate(force=True)
```

*However*, there is still an issue mentioned in #3915 and discussed more extensively in #2911 that the uninstall hooks take `Package` objects when they really should take `Spec`s.  The difference being that `Package` objects require an *instance* of a specific package, while `Spec`s are just descriptors.  So, the hook above won't be executed for packages that are no longer known to Spack, which is a bug.

I'm planning on reworking the hooks to take Specs to fix this and #2911.  Hooks will then need to properly think about what happens if the package they're dealing with no longer exists.

**Comment by svenevs** [comment]
Oh my, how deep the rabbit hole goes!  I see, well I'll close this since it's in queue elsewhere, but the above was going on even when `flann` was still a valid package.  So it's not just unknown packages, but uninstalled packages.

The original title indicates the confusion.  I thought there was like some sort of pitchfork bifurcation going on here with `activate` somehow remembering how something got installed.  It turns out it was crashing from the previous installation being activated, but already uninstalled.
